File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    void backtrack(int open_brackets, int closed_brackets, int n, string& stk, vector<string>& out)
    {
        if ( open_brackets == closed_brackets and open_brackets == n)
        {
            out.push_back(stk);
            return;
        }
        if ( open_brackets < n)
        {
            stk += "(";
            backtrack(open_brackets + 1, closed_brackets, n, stk, out);
            stk.pop_back();
        }
        if ( closed_brackets < open_brackets)
        {
            stk += ")";
            backtrack(open_brackets, closed_brackets + 1, n, stk, out);
            stk.pop_back();
        }
    }
    vector<string> generateParenthesis(int n) {
        string stk;
        vector<string> out;
        backtrack(0, 0, n, stk, out);
        return out;
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.cs
public class Solution {
    List<string> result = new List<string>(); 
    public IList<string> GenerateParenthesis(int n) {
        Iterate(n, 0, 0, "");
        return result;
    }
    public void Iterate(int n, int l, int r, string curr)
    {
        if (curr.Length == 2 * n)
        {
            result.Add(curr);
            return;
        }
        if (l < n)
        {
            Iterate(n, l + 1, r, curr + "(");
        }
        if (r < l)
        {
            Iterate(n, l, r + 1, curr + ")");
        }
    }
}
